core/src/main/java/com/klaytn/caver/wallet/keyring/Keyring.java

```diff
@@ -0,0 +1,741 @@
+     * @return Keyring
+     * @throws CipherException It throws when cipher operation has failed.
+     */
+    public static Keyring decrypt(KeyStore keystore, String password) throws CipherException{
+        if(keystore.getVersion() == KeyStore.KEY_VERSION_V3 && keystore.getCrypto() == null) {
+            throw new IllegalArgumentException("Invalid keystore V3 format: 'crypto' is not defined.");
+        } else if(keystore.getVersion() == KeyStore.KEY_VERSION_V4 && keystore.getKeyring() == null) {
+            throw new IllegalArgumentException("Invalid keystore V4 format: 'keyring' is not defined.");
+        }
+
+        if(keystore.getCrypto() != null) {
+            if(keystore.getKeyring() != null) {
+                throw new IllegalArgumentException("Invalid key store format: 'crypto' and 'keyring' cannot be defined together.");
+            }
+        }
+
+        if(keystore.getVersion() == KeyStore.KEY_VERSION_V3) {
+            KeyStore.Crypto crypto = keystore.getCrypto();
+            String privateKey = KeyStore.Crypto.decryptCrypto(crypto, password);
+            return Keyring.create(keystore.getAddress(), privateKey);
+        }
+
+        List keyring = keystore.getKeyring();
+        List<String[]> privateKeyList = new ArrayList<>();
+        if(keyring.get(0) instanceof KeyStore.Crypto) {
+            String[] privateKeyArr = new String[keyring.size()];
+            for(int i=0; i<keyring.size(); i++) {
+                KeyStore.Crypto crypto = (KeyStore.Crypto)keyring.get(i);
+                String privateKey = KeyStore.Crypto.decryptCrypto(crypto, password);
+                privateKeyArr[i] = privateKey;
+            }
+            privateKeyList.add(privateKeyArr);
+        } else {
+            for(List<KeyStore.Crypto> multiKeying : (List<List<KeyStore.Crypto>>)keyring) {
+                String[] privateKeyArr = new String[multiKeying.size()];
+                for(int i=0; i<multiKeying.size(); i++) {
+                    KeyStore.Crypto crypto = multiKeying.get(i);
+                    String privateKey = KeyStore.Crypto.decryptCrypto(crypto, password);
+                    privateKeyArr[i] = privateKey;
+                }
+                privateKeyList.add(privateKeyArr);
+            }
+        }
+
+        return Keyring.createWithRoleBasedKey(keystore.getAddress(), privateKeyList);
+    }
+
+    /**
+     * Recovers the address that was used to sign the given data.
+     * This function automatically creates a message hash by appending a Klaytn sign prefix to the message.
+     * @param messageSigned A MessageSigned object
+     * @return String
+     * @throws SignatureException It throws when recover operation has failed.
+     */
+    public static String recover(MessageSigned messageSigned) throws SignatureException {
+        KlaySignatureData klaySignatureData = messageSigned.getSignatureData();
+        return recover(messageSigned.getMessage(), klaySignatureData);
```

Comment:
```suggestion
        return recover(messageSigned. getMessageHash(), klaySignatureData, true);
```
We can get messageHash from MessageSigned, so how about calling recover(String, KlaySignatureData, boolean) directly?